HeterogeneousCore/AlpakaCore/README.md

```diff
@@ -0,0 +1,403 @@
+    * Use `stream::EDProducer`
+  * If you need to transfer some data back to host, use `stream::SynchronizingEDroducer`
+* All code using `ALPAKA_ACCELERATOR_NAMESPACE` should be placed in `Package/SubPackage/{interface,src,plugins,test}/alpaka` directory
+  * Alpaka-dependent code that uses templates instead of the namespace macro can be placed in `Package/SubPackage/interface` directory
+* All source files (not headers) using Alpaka device code (such as kernel call, functions called by kernels) must have a suffic `.dev.cc`, and be placed in the aforementioned `alpaka` subdirectory
+* Any code that `#include`s a header from the framework or from the `HeterogeneousCore/AlpakaCore`(?) must be separated from the Alpaka device code, and have the usual `.cc` suffix.
+  * A few framework headers are allowed to be used in `.dev.cc` files, e.g. `FWCore/Utilities/interface/Exception.h` or `FWCore/MessageLogger/interface/MessageLogger.h` (?)
+
+## Data formats
+
+Data formats, for both Event and EventSetup, should be placed following their usual rules. The Alpaka-specific conventions are
+* There must be a host-only flavor of the data format that is either idependent of Alpaka, or depends only on Alpaka's Serial backend
+  * The host-only data format must be defined in `Package/SubPackage/interface/` directory
+  * If the data format is to be serialized (with ROOT), it must be serialized in a way that the on-disk format does not depend on Alpaka, i.e. it can be read without Alpaka
+  * For Event data products the ROOT dictionary should be defined in `Package/SubPackage/src/classes{.h,_def.xml}` (**NOTE**: currently they are in `.../src/alpaka/classes_serial{.h,_def.xml}`
+    * As usual, the `classes_def.xml` should declare the dictionaries for the data product type `T` and `edm::Wrapper<T>`. These data products can be declared as persistent (default) or transient (`persistent="false"` attribute).
+  * For EventSetup data products [the registration macro `TYPELOOKUP_DATA_REG`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToRegisterESData) should be placed in `Package/SubPackage/src/ES_<type name>.cc`.
+* The device-side data formats are defined in `Package/SubPackage/interface/alpaka/` directory
+  * The device-side data format classes should be either templated over the device type, or defined in the `ALPAKA_ACCELERATOR_NAMESPACE` namespace.
+  * For Event data products the ROOT dictionary should be defined in `Package/SubPackage/src/alpaka/classes_<backend>{.h,_def.xml}`
+    * The `classes_<backend>_def.xml` should declare the dictionaries for the data product type `T`, `edm::DeviceProduct<T>`, and `edm::Wrapper<edm::DeviceProduct<T>>`. All these dictionaries must be declared as transient with `persistent="false"` attribute.
+    * The list of `<backend>` includes currently: `cuda`
```

Comment:
Maybe `<platform>` would make more sense for data formats than `<backend>`? (even if technically we have only "backend" concept right now)

Reply:
`<platform>` (well, `Platform`) is already an Alpaka concept, and indeed it seems to be a better match than `<backend>` here: Alpaka has platforms for "CPU", "CUDA", "HIP" (ROCm), _etc._, and doesn't differentiate between "CPU with serial backend" and "CPU with TBB parallel backend".

OpenMP 5, OpenACC and SYCL are their own platforms - but, to be fair, they likely require their own memory management.